core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/kopt/KOptListMoveSelectorFactory.java

```diff
@@ -0,0 +1,68 @@
+import org.optaplanner.core.impl.domain.variable.descriptor.GenuineVariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.descriptor.ListVariableDescriptor;
+import org.optaplanner.core.impl.heuristic.HeuristicConfigPolicy;
+import org.optaplanner.core.impl.heuristic.selector.entity.EntitySelector;
+import org.optaplanner.core.impl.heuristic.selector.entity.EntitySelectorFactory;
+import org.optaplanner.core.impl.heuristic.selector.move.AbstractMoveSelectorFactory;
+import org.optaplanner.core.impl.heuristic.selector.move.MoveSelector;
+import org.optaplanner.core.impl.heuristic.selector.value.ValueSelector;
+import org.optaplanner.core.impl.heuristic.selector.value.ValueSelectorFactory;
+
+public class KOptListMoveSelectorFactory<Solution_>
+        extends AbstractMoveSelectorFactory<Solution_, KOptListMoveSelectorConfig> {
+
+    private static final int DEFAULT_MINIMUM_K = 2;
+    private static final int DEFAULT_MAXIMUM_K = 4;
+
+    public KOptListMoveSelectorFactory(KOptListMoveSelectorConfig moveSelectorConfig) {
+        super(moveSelectorConfig);
+    }
+
+    @Override
+    protected MoveSelector<Solution_> buildBaseMoveSelector(HeuristicConfigPolicy<Solution_> configPolicy,
+            SelectionCacheType minimumCacheType, boolean randomSelection) {
+        SelectionOrder selectionOrder = SelectionOrder.fromRandomSelectionBoolean(randomSelection);
+        EntitySelectorConfig entitySelectorConfig = new EntitySelectorConfig();
+        // TODO let's not forget this here
+        //entitySelectorConfig.setCacheType(SelectionCacheType.PHASE);
+        EntitySelector<Solution_> entitySelector =
+                EntitySelectorFactory.<Solution_> create(entitySelectorConfig)
+                        .buildEntitySelector(configPolicy, minimumCacheType, selectionOrder);
+        // TODO support coexistence of list and basic variables https://issues.redhat.com/browse/PLANNER-2755
+        GenuineVariableDescriptor<Solution_> variableDescriptor =
+                getTheOnlyVariableDescriptor(entitySelector.getEntityDescriptor());
+        if (!variableDescriptor.isListVariable()) {
+            throw new IllegalArgumentException("The subListSwapMoveSelector (" + config
+                    + ") can only be used when the domain model has a list variable."
+                    + " Check your @" + PlanningEntity.class.getSimpleName()
+                    + " and make sure it has a @" + PlanningListVariable.class.getSimpleName() + ".");
+        }
+
+        ValueSelector<Solution_> valueSelector = buildEntityDependentValueSelector(configPolicy,
+                entitySelector.getEntityDescriptor(), minimumCacheType, selectionOrder);
+        int minimumK = Objects.requireNonNullElse(config.getMinimumK(), DEFAULT_MINIMUM_K);
+        int maximumK = Objects.requireNonNullElse(config.getMaximumK(), DEFAULT_MAXIMUM_K);
+        return new KOptListMoveSelector<>(((ListVariableDescriptor<Solution_>) variableDescriptor), entitySelector,
```

Comment:
I'm thinking... This algorithm gets k-opt done via a series of 2-opt moves. What if the user requests 2-opt only? 

Is there some sort of an optimization we could do? Bypassing all of the k-opt logic, going straight to the underlying 2-opt move?

Reply:
That is already done (and why TwoOptListMove exists); there would be minimal gains (the only things that can be removed is generating a "random" k and the check if `k == 2`